File: CppPerformanceDemo/fibonacci_calc.cpp
// fibonacci_calc.cpp
// To compile, run one of the following commands in your terminal:
// run with: make run
#include <iostream>
#include <chrono>

// Function to calculate the Nth Fibonacci number recursively
// This function is intentionally inefficient for demonstration purposes,
// as it recalculates the same values many times.
long long fibonacci(int n) {
    if (n <= 1) {
        return n;
    }
    return fibonacci(n - 1) + fibonacci(n - 2);
}

int main() {
    // Define the input value for the calculation (adjust as needed for performance)
    const int N = 40; 

    std::cout << "--- C++ Performance Demonstration ---" << std::endl;
    std::cout << "Calculating the " << N << "th Fibonacci number recursively..." << std::endl;

    // Start high-resolution timer
    auto start = std::chrono::high_resolution_clock::now();

    // Perform the calculation
    long long result = fibonacci(N);

    // Stop timer
    auto stop = std::chrono::high_resolution_clock::now();

    // Calculate the duration
    auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(stop - start);

    std::cout << "Result (Fib(" << N << ")): " << result << std::endl;
    std::cout << "Execution Time: " << duration.count() << " milliseconds" << std::endl;
    std::cout << "-----------------------------------" << std::endl;

    return 0;
}

File: NodeJsFlexibilityDemo/fibonacci_calc.ts
// fibonacci_calc.ts
// run with: npx tsx fibonacci_calc.ts
// The function is identical in logic to the C++ version
function fibonacci(n: number): number {
  if (n <= 1) {
    return n;
  }
  // NOTE: TypeScript/JavaScript uses standard numbers (doubles),
  // which might handle larger numbers differently than C++'s long long,
  // but the core computational load remains the same for comparison.
  return fibonacci(n - 1) + fibonacci(n - 2);
}
// Define the input value (must be the SAME as the C++ demo for fairness)
const N = 40;
function runCalculation() {
  console.log("--- Node.js Performance Demonstration ---");
  console.log(`Calculating the ${N}th Fibonacci number recursively...`);
  // Start timer
  const start = performance.now();
  // Perform the calculation
  const result = fibonacci(N);
  // Stop timer and calculate duration
  const end = performance.now();
  const duration = end - start;
  console.log(`Result (Fib(${N})): ${result}`);
  console.log(`Execution Time: ${duration.toFixed(2)} milliseconds`);
  console.log("-----------------------------------------");
  console.log(
    `\n(This timing will be significantly slower than the C++ version, which is the point of the demo.)`,
  );
}

runCalculation();
